### DISCUSSION TITLE
Proposal: Add possibility to inject cancellation token into while/for/foreach loops

### DISCUSSION THREAD

**Discussion by Kamil-Zakiev** [dotnet/csharplang#6505] [Language Ideas]
I just came up with an idea that is sounds pretty reasonable imho
I was thinking of how I can handle cancellation of a cpu-bounded operation.
Let's say it's a while loop over big input collection.
In code that would be

```csharp
while (someCondition)
{
      cancellationToken.ThrowIfCancellationRequested();
      
      // do some operation
}
```

But what if loops in C# would have supported cancellation out-of-the-box?
So the code would look like
```csharp
while (someCondition) considering cancellationToken
{
      // do some operation
}
for (var i = 0; i < 10; i++) considering cancellationToken
{
      // do some operation
}
```

What do you think? would it be cool or cumbersome? waiting for your feedback :)

**Comment by CyrusNajmabadi** [comment]
Seems to be virtually identical to the existing pattern for me.  both effectively hae the cancellation check, just one is before the `{` and one is after it.  :)

**Comment by Richiban** [comment]
It would be nicer, IMO if C# could get some kind of inline method and inline lambda parameter support, which would allow you to write your own loops that would have no overhead (for examples see Kotlin's inline support: https://kotlinlang.org/docs/inline-functions.html).

This would allow you to write your own method:

```cs
public inline void WhileNotCancelled(CancellationToken ct, Function<bool> condition, Action body)
{
	while (condition())
	{
		cancellationToken.ThrowIfCancellationRequested();

		body();
	}
}
```

and you could use it like this:

```cs
WhileNotCancelled(ct, () => condition, () => // do some operation);
```

and there would be no overhead. Granted, C# doesn't have the syntax support for lambdas to make this look nicer, but it would work.
